### ISSUE TITLE
Fetch API corrupts repetitive response headers

### ISSUE THREAD

**Issue by sshymko** [facebook/react-native#12846]
### Description

The [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) misinterprets repetitive response headers, for instance, `Set-Cookie` header passed separately for each cookie. All values of the header concatenate into a single string unusable for further processing.

### Reproduction

```js
fetch('https://github.com/facebook/react-native')
    .then((response) => {
        console.log(response.headers);
        console.log(response.headers.getAll('Set-Cookie'));
    });
```

**Expected result:**
```js
{ map: 
	   { date: [ 'Fri, 10 Mar 2017 06:53:22 GMT' ],
	     'content-type': [ 'text/html; charset=utf-8' ],
	     'set-cookie': [
                 '_gh_sess=eyJsYXN0...; path=/; secure; HttpOnly',
                 'user_session=AYIZ4afT...; path=/; expires=Fri, 24 Mar 2017 07:18:28 -0000; secure; HttpOnly'
             ]
           }
}
```
```js
[
    '_gh_sess=eyJsYXN0...; path=/; secure; HttpOnly',
    'user_session=AYIZ4afT...; path=/; expires=Fri, 24 Mar 2017 07:18:28 -0000; secure; HttpOnly'
]
```

**Actual result:**
```js
{ map: 
	   { date: [ 'Fri, 10 Mar 2017 06:53:22 GMT' ],
	     'content-type': [ 'text/html; charset=utf-8' ],
	     'set-cookie': [
                 '_gh_sess=eyJsYXN0...; path=/; secure; HttpOnly, user_session=AYIZ4afT...; path=/; expires=Fri, 24 Mar 2017 07:18:28 -0000; secure; HttpOnly'
             ]
           }
}
```
```js
[
    '_gh_sess=eyJsYXN0...; path=/; secure; HttpOnly, user_session=AYIZ4afT...; path=/; expires=Fri, 24 Mar 2017 07:18:28 -0000; secure; HttpOnly'
]
```

### Solution

The underlying platform-specific code that implements the Fetch API needs to be fixed.

### Additional Information

* React Native version: 0.42.0
* Platform: iOS (haven't tested on Android)
* Operating System: MacOS

**Comment by sshymko** [comment]
React Native also appears to discard cookie path (and perhaps domain) when automatically sending `Cookie` headers based on the previously received `Set-Cookie`. It may be unable to recognize cookie properties because values of multiple cookies are "glued" in the underlying data structure.

**Comment by nfriedly** [comment]
In reading things this morning, and [talking to folks behind the Fetch spec](https://github.com/whatwg/fetch/issues/506), React Native seems to be following the [fetch spec](https://fetch.spec.whatwg.org/#headers-class) correctly, with the exception of allowing access to `Set-Cookie` headers at all. 

The header combining behavior is allowed (but not required) according to [the HTTP RFC](https://tools.ietf.org/html/rfc2616#page-31), and also mentioned in the cookie RFC:

> An origin server may include multiple Set-Cookie headers in a response.  Note that an intervening gateway could fold multiple such headers into a single header.

It can be rather difficult to combine a combined Set-Cookie header since, for example, a expires field can contain the same "`, `" separator that is used to combine multiple headers into one.

The Fetch spec takes things a step further and requires this combining behavior. Fetch folks are aware that it doesn't work well for cookies but require the behavior anyways since their spec doesn't allow access to `Set-Cookie` headers.

So, in summary...

Since React Native does allow access to set-cookie headers, I would recommend following the [example of Node.js](https://tools.ietf.org/html/rfc2616#page-31) and special-casing `Set-Cookie` headers as an array field rather than a combined one and/or providing access to the raw headers. 

---
Regarding this:

> React Native also appears to discard cookie path (and perhaps domain) when automatically sending Cookie headers based on the previously received Set-Cookie. 

I believe this is the correct behavior. `Set-Cookie` is allowed to contain that stuff, but not `Cookie` - it just contains name-value pairs as appropriate for the requested path and domain.

**Comment by sshymko** [comment]
Just wanted to clarify regarding:
> I believe this is the correct behavior. `Set-Cookie` is allowed to contain that stuff, but not `Cookie` - it just contains name-value pairs as appropriate for the requested path and domain.

React Native's `fetch` acts as a browser in terms of maintaining cookies. When a response contains `Set-Cookie` header, React Native memorizes it and starts to include the respective `Cookie` header into all subsequent requests. That is pretty cool, isn't it? Well, the problem is that it does not respect the cookie scope (path for sure, haven't checked the domain). For example, if a request to http://login.example.com/ returns `Set-Cookie: token=123; Path=/protected/`, React Native will keep sending `Cookie: token=123` to URL with any path, not only the `/protected/` path the cookie was meant for. This a security concern.

**Comment by bitinn** [comment]
I am the owner of `node-fetch` (which react-native uses) and was searching for a relevant issue about `getAll`.

Basically there are 2 issues to be fixed for `node-fetch` here:

- for `v1.6.3`, which `react-native` currently uses, you are saying `set-cookie` headers isn't parsed properly. I would glad to take a PR with tests to fix it.

- we are currently working on `v2` of `node-fetch`, which follows the Fetch Spec more closely, but because Fetch Spec didn't account for `set-cookie` headers being exposed, and has deprecated `getAll`, we somehow need to handle this special case which is not possible in browsers. I would glad to take a PR with tests as well.

As for the cookie path issue, I think its a `react-native` problem, true?

You can follow up on a similar issue on https://github.com/bitinn/node-fetch/issues/251

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while. Maybe the issue has been fixed in a recent release, or perhaps it is not affecting a lot of people. Either way, we're automatically closing issues after a period of inactivity. Please do not take it personally!

If you think this issue should definitely remain open, please let us know. The following information is helpful when it comes to determining if the issue should be re-opened:

- Does the issue still reproduce on the latest release candidate? Post a comment with the version you tested.
- If so, is there any information missing from the bug report? Post a comment with all the information required by the [issue template](https://github.com/facebook/react-native/blob/master/.github/ISSUE_TEMPLATE.md).
- Is there a pull request that addresses this issue? Post a comment with the PR number so we can follow up.

If you would like to work on a patch to fix the issue, *contributions are very welcome*! Read through the [contribution guide](http://facebook.github.io/react-native/docs/contributing.html), and feel free to hop into [#react-native](https://discordapp.com/invite/0ZcbPKXt5bZjGY5n) if you need help planning your contribution.

**Comment by nfriedly** [comment]
I know this ticket is pretty old, but I think I finally have a good solution. Thanks to a [PR](https://github.com/nfriedly/set-cookie-parser/pull/19) from @chrusart, the `set-cookie-parser` library is now capable of splitting the combined cookie string! It can then be parsed as individual set-cookie headers. See https://www.npmjs.com/package/set-cookie-parser#usage-in-react-native

Please test it out and let me know how it works for you.
